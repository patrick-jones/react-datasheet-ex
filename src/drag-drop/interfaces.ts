import RDS from 'react-datasheet';
import {ConnectDragPreview, ConnectDragSource, ConnectDropTarget} from 'react-dnd';
import {ColumnHeader, HeaderRendererProps, RowRendererProps} from '../interfaces';

/**
 * Properties that must be provided to a drag and drop enabled row
 */
export interface DragDropRowProps<T extends RDS.Cell<T, V>, V = string> {
  /** The current row index */
  row: number;
  /** Callback for when a row is dropped on another */
  readonly onRowDropped: (source: number, target: number) => void;
}

/**
 * Properties that must be provided to a drag and drop enabled column header
 */
export interface DragDropHeaderProps {
  /** The current column index */
  col: number;
  /** Callback for one header is dropped on another */
  readonly onHeaderDropped: (source: number, target: number) => void;
}

/**
 * Drag source properties that are injected into a drag and drop provider
 */
export interface InjectedDragSourceProps {
  readonly connectDragSource: ConnectDragSource;
  readonly connectDragPreview: ConnectDragPreview;
}

/**
 * Drop target properties that are injected into a drag and drop provider
 */
export interface InjectedDropTargetProps {
  readonly connectDropTarget: ConnectDropTarget;
  readonly isDropOver: boolean;
}

/**
 * Combined drag source and drop target properties that are injected
 * into a drag and drop provider
 */
export interface InjectedDragDropProps extends InjectedDragSourceProps, InjectedDropTargetProps {}

/**
 * Extends `ReactDatasheet.RowRendererProps` to include
 * drag and drop properties
 */
export interface DragDropRowRendererProps<CellType, V = string>
  extends RowRendererProps<CellType, V>,
    DragDropRowProps<CellType, V> {}

/**
 * Extends `HeaderRendererProps` to include drag and drop properties.
 */
export interface DragDropHeaderRendererProps<T extends ColumnHeader = ColumnHeader>
  extends HeaderRendererProps<T>,
    DragDropHeaderProps {}
