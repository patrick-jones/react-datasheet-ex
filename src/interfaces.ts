import React, {CSSProperties} from 'react';
import RDS from 'react-datasheet';

export interface RowLocation {
  row: number;
}

export interface ColumnLocation {
  col: number;
}

export interface ColumnHeader {
  id: string;
  title: string;
  width?: number;
}

export interface HeaderRendererProps<T extends ColumnHeader = ColumnHeader> extends ColumnLocation {
  header: T;
}

export type HeaderRenderer<T extends ColumnHeader = ColumnHeader,
  R extends HeaderRendererProps<T> = HeaderRendererProps<T>> = React.ComponentType<R>;

export type HeaderContentRenderer<T extends ColumnHeader = ColumnHeader>
  = React.ComponentType<HeaderRendererProps<T>>;

export interface SheetRendererProps<
    T extends RDS.Cell<T, V>, V = string, H extends ColumnHeader = ColumnHeader
  >
  extends RDS.SheetRendererProps<T, V> {
  /** React style object that should be applied to the enclosing table */
  style?: CSSProperties | undefined;
  /**
   * An array of `ColumnHeader`, which are rendered into `th` elements
   * @default []
   */
  headers: H[];
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderRenderer
   */
  headerRenderer: HeaderRenderer;
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderContentRenderer
   */
  headerContentRenderer: HeaderContentRenderer;
}


export type ActionContentRenderer<T extends RowLocation = RowLocation>
  = React.ComponentType<T>;

export interface RowRendererProps<
    T extends RDS.Cell<T, V>, V = string, AP extends RowLocation = RowLocation
  > extends RDS.RowRendererProps<T, V> {

  actionContentRenderer?: ActionContentRenderer<AP>;
}
