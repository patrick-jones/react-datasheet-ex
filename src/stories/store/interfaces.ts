import RDS from 'react-datasheet';
import {ColumnHeader} from '../../interfaces';

export type ExampleCellValue = string | number | null;
export interface ExampleCellType extends RDS.Cell<ExampleCellType> {
  value: ExampleCellValue;
}

export type ExampleModel = {
  id: string;
  headers: ColumnHeader[];
  rows: ExampleRow[];
  selected: RDS.Selection | null;
};

export type ExampleRow = {
  id: string;
  selected?: boolean;
  data: ExampleCellType[];
};

export type Grid = ExampleCellType[][];
