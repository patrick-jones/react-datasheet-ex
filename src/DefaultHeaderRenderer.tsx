import React, {Component, ReactNode} from 'react';
import {ColumnHeader, HeaderRendererProps} from './interfaces';


export default class DefaultHeaderRenderer<T extends ColumnHeader = ColumnHeader>
  extends Component<HeaderRendererProps<T>> {

  render(): ReactNode {
    const {header: {title, width, id}, children} = this.props;
    const style = typeof width === 'undefined' ? {} : {width: `${width}px`};
    return (
      <th
        key={id}
        title={title}
        style={style}
        className='cell read-only rdr-sheet-renderer__header-cell'
      >
        {children}
      </th>
    );
  }
}
