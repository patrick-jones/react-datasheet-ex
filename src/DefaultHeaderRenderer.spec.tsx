import React from 'react';
import {shallow} from 'enzyme';
import DefaultHeaderRenderer from './DefaultHeaderRenderer';
import {ColumnHeader} from './interfaces';

describe('DefaultHeaderRenderer', () => {
  const header: ColumnHeader = {
    id: 'foo',
    title: 'AA',
  };

  it('Renders text as expected', () => {
    const wrapper = shallow(
      <DefaultHeaderRenderer col={0} header={header}>{header.id}</DefaultHeaderRenderer>
    );
    expect(wrapper.find('th.rdr-sheet-renderer__header-cell').text()).toBe(header.id);
  });

  it('Renders style as needed', () => {
    const wrapper = shallow(
      <DefaultHeaderRenderer col={0} header={header}>{header.id}</DefaultHeaderRenderer>
    );
    expect(wrapper.find('th.rdr-sheet-renderer__header-cell').prop('style')).toEqual({});

    const styled = shallow(
      <DefaultHeaderRenderer col={0} header={{...header, width: 100}}>
        {header.id}
      </DefaultHeaderRenderer>
    );
    expect(styled.find('th.rdr-sheet-renderer__header-cell').prop('style'))
      .toEqual({width: '100px'});
  });

  it('matches snapshot', () => {
    const wrapper = shallow(
      <DefaultHeaderRenderer col={0} header={header}>{header.id}</DefaultHeaderRenderer>
    );
    expect(wrapper).toMatchSnapshot();
  });

});
