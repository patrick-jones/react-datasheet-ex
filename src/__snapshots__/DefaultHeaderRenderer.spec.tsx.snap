// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultHeaderRenderer matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DefaultHeaderRenderer
    col={0}
    header={
      Object {
        "id": "foo",
        "title": "AA",
      }
    }
  >
    foo
  </DefaultHeaderRenderer>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "foo",
    "nodeType": "host",
    "props": Object {
      "children": "foo",
      "className": "cell read-only rdr-sheet-renderer__header-cell",
      "style": Object {},
      "title": "AA",
    },
    "ref": null,
    "rendered": "foo",
    "type": "th",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "foo",
      "nodeType": "host",
      "props": Object {
        "children": "foo",
        "className": "cell read-only rdr-sheet-renderer__header-cell",
        "style": Object {},
        "title": "AA",
      },
      "ref": null,
      "rendered": "foo",
      "type": "th",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
